patchwork-extensions-keybinds/src/main/java/com/patchworkmc/mixin/extensions/keybinds/MixinKeyBinding.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2020, 2019-2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package com.patchworkmc.mixin.extensions.keybinds;
+
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+
+import net.minecraftforge.client.extensions.IForgeKeybinding;
+import net.minecraftforge.client.settings.IKeyConflictContext;
+import net.minecraftforge.client.settings.KeyBindingMap;
+import net.minecraftforge.client.settings.KeyConflictContext;
+import net.minecraftforge.client.settings.KeyModifier;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Mutable;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.client.options.KeyBinding;
+import net.minecraft.client.resource.language.I18n;
+import net.minecraft.client.util.InputUtil;
+
+import com.patchworkmc.impl.extensions.keybinds.IPatchworkKeyBinding;
+
+@Mixin(KeyBinding.class)
+public abstract class MixinKeyBinding implements IForgeKeybinding, IPatchworkKeyBinding {
+	@Shadow
+	private InputUtil.KeyCode keyCode;
+
+	@Mutable
+	@Shadow
+	@Final
+	private static Map<InputUtil.KeyCode, KeyBinding> keysByCode;
+
+	@Shadow
+	public abstract String getCategory();
+
+	@Shadow
+	public abstract String getId();
+
+	@Shadow
+	@Final
+	private static Map<String, Integer> categoryOrderMap;
+	private KeyModifier keyModifierDefault = KeyModifier.NONE;
+	private KeyModifier keyModifier = KeyModifier.NONE;
+	private IKeyConflictContext keyConflictContext = KeyConflictContext.UNIVERSAL;
+
+	@Inject(method = "<clinit>", at = @At("RETURN"))
+	private static void staticInit(CallbackInfo ci) {
+		keysByCode = new KeyBindingMap();
```

Comment:
I dislike this approach of overwriting the vanilla map and behavior. Instead, it makes more sense to just have a secondary map for all of the non-NONE KeyModifiers alongside the vanilla map, and then in the methods that reference the vanilla map update the modifer map as well.

Reply:
Forge allows two keybinds to have the same keycode with this approach, vanilla is really bad at this.

Vanilla is just a map from KeyCode to KeyBinding.
```java
private static final Map<InputUtil.KeyCode, KeyBinding> keysByCode = Maps.newHashMap();
```
However, I don't think overwriting this is such a bad idea, this mimics the vanilla map function and anyone using the map would be compatible.

The next question is do we really have to support multiple keybinds with the same keycode, probably not, but I think it is kind of worth it if we are going to implement modifiers anyways, we are editing vanilla behaviour anyways.

Reply:
What I'm trying to say is have the maps for each other key modifier exist alongside the vanilla map, because effectively the vanilla map is already working as the map for the NONE keymodifier so replacing it is unnecessary.

Reply:
We still have to overwrite the methods if the user uses the modifers, and other mods using the map will simply ignore the modifiers.